keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md

```diff
@@ -0,0 +1,308 @@
+  not exist yet,
+* gets the list of participating API servers,
+* updates the storageVersion locally. Specifically,
+  * creates or updates the .status.serverStorageVersions, to express this API
+    server's decodableVersions and encodingVersion.
+  * removes .status.serverStorageVersions entries whose server ID is not present
+    in the list of participating API servers, such entries are stale.
+  * checks if all participating API servers agree on the same storage version.
+    If so, sets the version as the status.agreedEncodingVersion. If not, sets
+    the status.agreedEncodingVersion to "No Agreed Version".
+* updates the storageVersion object, using the rv in the first step
+  to avoid conflicting with other API servers.
+* installs the resource handler.
+
+### Garbage collection
+
+There are two kinds of "garbage":
+
+1. stale storageVersion.status.serverStorageVersions entries left by API servers
+   that have gone away;
+2. storageVersion objects for resources that are no longer served.
+
+We can't rely on API servers to remove the first kind of stale entries during
+bootstrap, because an API server can go away after other API servers bootstrap,
+then its stale entries will remain in the system until one of the other API
+servers reboots.
+
+Hence, we propose a leader-elected control loop in API server to clean up the
+stale entries. The control loop watches the list of participating API servers,
+upon changes, it performs the following actions for each storageVersion object:
+
+* gets a storageVersion object
+* gets the list of participating API servers,
+* locally, removes the stale entries in
+  storageVersion.status.serverStorageVersions,
+  * after the removal, if all participating API servers have the same
+    encodingVersion, then sets storageVersion.status.AgreedEncodingVersion. 
+* checks if the storageVersion.status.serverStorageVersions is empty,
+  * if empty, deletes the storageVersion object,
+  * otherwise updates the storageVersion object,
+  * both the delete and update operations are preconditioned with the rv in the
+    first step to avoid conflicting with API servers modifying the object.
+
+An API server needs to establish its membership in the list of participating API
+servers before updating storageVersion, otherwise the above control loop can
+mistake a storageVersion.status.serverStorageVersions entry added by a new API
+server as a stale entry.
+
+### CRDs
+
+Today, the [storageVersionHash][] in the discovery document in HA setup can
+diverge from the actual storage version being used. See the [appendix][] for
+details.
+
+[appendix]:#appendix
+[storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/c008cf95a92c5bbea67aeab6a765d7cb1ac68bd7/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L989
+
+To accurately reflect the storage version being used, the apiextension-apiserver
+needs to update the storageVersion object when it [creates][] the custom
+resource handler.
```

Comment:
can crd storage version change dynamically, after the custom resource handler is created?